### PR TITLE
[Debugger] Add support to configure environmental variables in launch mode

### PR DESCRIPTION
**Pull request by NipunaRanasinghe**
## Purpose
This PR
1. improves debug server client configuration implementation and 
2. adds support for environmental variables, which can be configured from the client (vscode plugin) side.
Screenshot from 2021-02-23 15-23-13

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/18608 - Cannot pass environment variables to debug tests
Fixes https://github.com/ballerina-platform/plugin-vscode/issues/3 - Support envFile / envorinment tags for vscode launch configurations

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [x] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java [hunk 1]
```diff
@@ -210,68 +216,43 @@ public void setExecutionManager(DebugExecutionManager executionManager) {
...
+            String hostName = configHolder.getHostName().orElse("");
+            int portName = configHolder.getDebuggePort();
+            attachToRemoteVM(hostName, portName);
+        } catch (IOException | IllegalConnectorArgumentsException | ClientConfigurationException e) {
+            String host = ((ClientAttachConfigHolder) clientConfigHolder).getHostName().orElse(LOCAL_HOST);
+            String portName;
+            try {
+                portName = Integer.toString(clientConfigHolder.getDebuggePort());
+            } catch (ClientConfigurationException clientConfigurationException) {
+                portName = "unknown";
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:253]: Shall we move `unknown` to a constant?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java [hunk 2]
```diff
@@ -895,6 +875,77 @@ private static boolean isValidFrame(StackFrame stackFrame) {
...
+                }
+            } catch (IOException ignored) {
+                // no-op
+            } catch (ClientConfigurationException | IllegalConnectorArgumentsException e) {
+                String host = ((ClientAttachConfigHolder) clientConfigHolder).getHostName().orElse(LOCAL_HOST);
+                String portName;
+                try {
+                    portName = Integer.toString(clientConfigHolder.getDebuggePort());
+                } catch (ClientConfigurationException clientConfigurationException) {
+                    portName = "unknown";
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java:923]: Same [here](https://github.com/ballerina-platform/ballerina-lang/pull/28842/files#r581564302)

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientConfigHolder.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    public int getDebuggePort() throws ClientConfigurationException {
+        if (debuggePort == null) {
+            failIfMissing(ARG_DEBUGGEE_PORT);
+            debuggePort = Integer.parseInt(clientRequestArgs.get(ARG_DEBUGGEE_PORT).toString());
+        }
+        return debuggePort;
+    }
+
+    protected void failIfMissing(String configName) throws ClientConfigurationException {
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientConfigHolder.java:62]: Shall we use a meaningful name here? `failIfConfigMissing` or `failIfClientConfigMissing`

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientConfigKind.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Defines the possible types of client(user) configurations.
+ *
+ * @since 2.0.0
+ */
+public enum ClientConfigKind {
+    ATTACH_CONFIG,
+    LAUNCH_CONFIG
+}
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientConfigKind.java:27]: Do we need a separate class here only for `ClientConfigKind`?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientLaunchConfigHolder.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            } else if (testDebugObj instanceof String) {
+                testDebug = Boolean.parseBoolean((String) testDebugObj);
+            } else {
+                testDebug = false;
+            }
+        }
+        return testDebug;
+    }
+
+    public boolean networkLogsEnabled() {
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientLaunchConfigHolder.java:101]: How about changing the name to `isNetworkLogsEnabled` here?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://wso2.com) All Rights Reserved.
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java:2]:
```suggestion
 * Copyright (c) 2021, WSO2 Inc. (http://wso2.com) All Rights Reserved.
```

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * Abstract implementation of Ballerina program launcher.
+ */
+public abstract class ProgramLauncher {
+
+    protected final String projectRoot;
+    protected final ClientLaunchConfigHolder configHolder;
+
+    private static final String BAL_RUN_CMD_NAME = "run";
+    private static final String BAL_TEST_CMD_NAME = "test";
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java:37]: How about changing the name to `BAL_RUN_CMD` and `BAL_TEST_CMD` here?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java (hunk 6 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java:2]: Only renamed the class name. We dont need to change the copyright year right?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 1 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java]: +1

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java (hunk 2 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/JBallerinaDebugServer.java]: Done

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientConfigHolder.java (hunk 3 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientConfigHolder.java]: changed to `failIfConfigMissing`

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientConfigKind.java (hunk 4 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientConfigKind.java]: Moved to `ClientConfigHolder`

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientLaunchConfigHolder.java (hunk 5 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/config/ClientLaunchConfigHolder.java]: Renamed

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java (hunk 7 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java]: +1 changed

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java (hunk 6 above)
Comment (praveennadarajah) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/launch/ProgramLauncher.java:2]: Then no need to change it. PR looks like you added a new class

**Review by praveennadarajah — APPROVED** [review]

**Pull request merged**